swiftpkg/bzlmod/swift_deps.bzl

```diff
@@ -183,11 +195,86 @@ def _declare_pkg_from_dependency(dep, config_pkg):
             dependencies_index = None,
         )
 
+def _declare_swift_package_repo(name, from_package, config_swift_package):
+    config_swift_package_kwargs = {}
+    for k in SWIFT_PACKAGE_CONFIG_ATTRS:
+        v = getattr(config_swift_package, k, None)
+        if v != None:
+            config_swift_package_kwargs[k] = v
+
+    _swift_package_repo(
+        name = name,
+        package = "{package}/{name}".format(
+            package = from_package.swift.package,
+            name = from_package.swift.name,
+        ),
+        **config_swift_package_kwargs,
+    )
+
+def _swift_package_repo_impl(repository_ctx):
+    package_path = repository_ctx.attr.package
+
+    # Construct the list of keyword arguments for the `swift_package_tool` rule.
+    kwargs = {}
+    for k in SWIFT_PACKAGE_CONFIG_ATTRS:
+        v = getattr(repository_ctx.attr, k, None)
+        if v != None:
+            kwargs[k] = v
+
+    # String should be "key = \"value\""
+    # NOTE: only supports string typed values as they are all quoted
+    kwarg_content = ",\n".join([
+        "    {key} = \"{value}\"".format(key = k, value = v)
+        for k, v in kwargs.items()
+    ])
+
+    repository_ctx.file(
+        "BUILD.bazel",
+        content = """
+load("@rules_swift_package_manager//swiftpkg:defs.bzl", "swift_package_tool")
+
+swift_package_tool(
+    name = "update",
+    cmd = "update",
+    package = "{package}",
+{kwarg_content}
+)
+
+swift_package_tool(
+    name = "resolve",
+    cmd = "resolve",
+    package = "{package}",
+{kwarg_content}
+)
+""".format(
+            package = package_path,
+            kwarg_content = kwarg_content,
+        ),
+    )
+
+_swift_package_repo = repository_rule(
```

Comment:
I think that we should move the repo rule to its own file under `swiftpkg/internal`. The other repository rules live there.

Reply:
I didnt know what to call this file since theres already a `swift_package.bzl` rule somewhere in there. Thoughts?

Reply:
Maybe `swift_package_tool_repo.bzl`?

Reply:
`swift_package_tool_repo.bzl` works. 

TBH, I almost commented that this functionality could stand to live in `rules_swift`. Right now, it is focused on Swift package management. However, in reality, we are just exposing the Swift binary. 🤔

Reply:
Extracted the repository rule to its own file let me know what you think

Reply:
Do agree though that a lot of this might also make sense in rules_swift. I think it's less obvious what it would be used for there though since Package.swift is strictly an SPM thing and the current rules don't really focus on that